### PR TITLE
Typography: Switch to ToolsPanel for block support UI

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Depends on:
- ~https://github.com/WordPress/gutenberg/pull/32392~
- ~https://github.com/WordPress/gutenberg/pull/34725~
- https://github.com/WordPress/gutenberg/pull/35415

Related: 
- https://github.com/WordPress/gutenberg/issues/27331
- https://github.com/WordPress/gutenberg/pull/33743
- https://github.com/WordPress/gutenberg/pull/34027

## Description

This PR:
- Adds new `typography` InspectorControls group slot
- Updates the typography block support hooks to provide callbacks required for ToolsPanel
- Updates the typography hook to render the typography support controls via the new InspectorControls slot

### Default Controls

_This PR only addresses the switch to the new `ToolsPanel` component. Deciding which controls should display by default for which blocks will be addressed via a separate PR._

## How has this been tested?

Manually.

#### Test instructions

1. Edit a post, add a navigation block, then select that block.
2. Confirm that the typography controls within the sidebar are within a ToolsPanel
3. Confirm each typography control works appropriately (Note: this PR does not make any specific block's typography controls "default" in terms of always displaying within the new panel)

## Screenshots 

https://user-images.githubusercontent.com/60436221/138071083-96932e63-5796-471b-90fd-44abec38da6a.mp4

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice one. In general for all these toolspanel refactors, we want to tread carefully with changing too much from what's shipping at the moment. _Ideally_ give or take a few exceptions, we want the default design tools to be the same as they were before the panel switch. The bigger benefit of the tools panel is for any _future_ features to benefit global styles and be available in a non-default way in the post editor. 

So in this case, we want to at the very least show font size and line height by default. What do you think?

**Comment by apeatling** [comment]
@jasmussen There's a bit of a mix of what blocks show for Typography by default. I think I agree that we should stick to font size and line height as the default for a lot of these unless there's a good reason not to?

Here's the blocks using Typography and what they currently show as the default. Let me know how we should move forward:

* **Button**
   -- Currently shows: Font family, Font size
* **Code**
   -- Currently shows: Font size
* **Heading**
   -- Currently shows: Font size, Font weight, Line height
* **List**
   -- Currently shows: Font family, Font size
* **Navigation**
   -- Currently shows: Font family, Font size, Appearance, Line height, Decoration, Letter case
* **Paragraph**
   -- Currently shows: Font size, Line height
* **Post Author**
   -- Currently shows: Font size, Line height
* **Post Comments**
   -- Currently shows: Font size, Line height
* **Post Comments Count**
   -- Currently shows: Font size, Line height
* **Post Comments Form**
   -- Currently shows: Font size, Line height
* **Post Comments Link**
   -- Currently shows: Font size, Line height
* **Post Date**
   -- Currently shows: Font size, Line height
* **Post Excerpt**
   -- Currently shows: Font size, Line height
* **Post Categories**
   -- Currently shows: Font size, Line height
* **Post Tags**
   -- Currently shows: Font size, Line height
* **Post Title**
   -- Currently shows: Font family, Font size, Line height
* **Preformatted**
   -- Currently shows: Font size
* **Query Pagination**
   -- Currently shows: Font size, Line height
* **Query Title**
   -- Currently shows: Font family, Font size, Line height
* **Site Tagline**
   -- Currently shows: Font family, Font size, Line height, Letter case
* **Site Title**
   -- Currently shows: Font family, Font size, Appearance, Line height, Decoration, Letter case
* **Term Description**
   -- Currently shows: Font size, Line height
* **Verse**
   -- Currently shows: Font family, Font size

**Comment by apeatling** [comment]
I think we could merge this work first, and then follow up on a decision on defaults in a separate PR.

**Comment by jasmussen** [comment]
If we were to merge this one with a completely empty typography panel for all those blocks, we'd want to move pretty fast to at least enable font size and lineheight for the Paragraph block. It honestly seems better to at least enable those two by default for all these as a starting point.

**Comment by apeatling** [comment]
> It honestly seems better to at least enable those two by default for all these as a starting point.

Agreed, if we stick to font size and line height as the two defaults for everything, then we could add it to this PR /cc @aaronrobertshaw

**Comment by aaronrobertshaw** [comment]
> Agreed, if we stick to font size and line height as the two defaults for everything, then we could add it to this PR /cc 

The block.json changes are easily done though I have two thoughts I'd like feedback on.

1. This PR already touches a lot of files. I think it would be best to have all the block.json in a separate PR. Is this ok? I can base it off this and have it ready to go so the two can be merged together.

2. Some of the blocks in the list above don't have line-height support enabled for them by the looks of it. Are we happy to opt into that support?

**Comment by ramonjd** [comment]
> I think it would be best to have all the block.json in a separate PR. 

There's a PR over here: https://github.com/WordPress/gutenberg/pull/34064 It's based off trunk for now.

 I can rebase it on top of this for testing once we're ready.

> Some of the blocks in the list above don't have line-height support enabled for them by the looks of it. Are we happy to opt into that support?

For now I've defaulted to which control was available, that is, setting fontSize and lineHeight defaults where each block supports them.

**Comment by jasmussen** [comment]
> Some of the blocks in the list above don't have line-height support enabled for them by the looks of it. Are we happy to opt into that support?

Is it not yet possible to tailor the support on a per-block basis?

**Comment by aaronrobertshaw** [comment]
> Is it not yet possible to tailor the support on a per-block basis?

It is. That's why some blocks have opted into the support while others haven't. We can also choose which controls are defaults on a per-block basis.

> It honestly seems better to at least enable those two by default for all these as a starting point.

I only wanted to confirm that we do indeed want to turn on line height for the blocks that hadn't already opted into it. Or, whether we wanted to just display font size and line height as default controls if the blocks currently support them. 

For example, the paragraph block has font size and line-height support, so it would show both by default. The code block only has font size support at the moment. Does it only show the font size control by default, or do we also update it to opt into line-height support and display that control as well?

**Comment by jasmussen** [comment]
> I only wanted to confirm that we do indeed want to turn on line height for the blocks that hadn't already opted into it. Or, whether we wanted to just display font size and line height as default controls if the blocks currently support them.

Right. Having both font size and line-height feel valuable, the only niggle is we want to make sure the line-height works well before enabling it. So for those that haven't yet had access to lineheight we might want to start without it and go from there.

**Comment by aaronrobertshaw** [comment]
This has been rebased onto #34725 which updates the new InspectorControls slots for block supports to use virtual bubbling. 

This PR has also been updated so the typography support hook injects controls via the typography slot instead of creating the ToolsPanel itself.

**Review by apeatling — COMMENTED** [review]
✅ Confirmed in the post editor all typography settings are working correctly with ToolsPanel:



✅ Confirmed that in the site editor and global styles panel all typography settings are displayed by default:



✅ Confirmed using the post title block that the typography settings appear by default in the site editor, and work as expected:

**Comment by aaronrobertshaw** [comment]
My apologies to all who were caught up in the mass review request after the previous push. Sorry!

File: packages/block-editor/src/hooks/font-appearance.js
Comment (ciampo) [inline packages/block-editor/src/hooks/font-appearance.js:41]: What is the reason behind removing these lines?

Same question for all of the `isDisabled` checks that have been removed (font-family, font-size, letter-spacing, line-height, text-decoration, text-transform, ...)

File: packages/block-editor/src/hooks/typography.scss
Comment (ciampo) [inline packages/block-editor/src/hooks/typography.scss]: If you feel that these are issues/limitations with the components that we're using, it'd be great if you could open individual issues for them and add them to tracking issue #34345

File: packages/components/src/font-size-picker/index.js
Comment (ciampo) [inline packages/components/src/font-size-picker/index.js]: Is `allowReset` a prop that is also used on other similar components?

If this is a new prop, we should add documentation for it (e.g. in the README) and add a CHANGELOG entry. It would be good if these changes would happen in a separate PR too (if it is too complicated to split the work at this point, at least we should keep this in mind for future work).

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [inline packages/components/src/tools-panel/styles.ts]: I understand if this is necessary at this point, but let's not forget that we should soon refactor away from hardcoded classnames.

**Review by ciampo — COMMENTED** [review]
Thank you for working on this, @aaronrobertshaw ! I left a few comments, and in general, I'll leave the testing of the overall functionality to other reviewers.

Mixing `@wordpress/components` changes into big PRs like this is not ideal, I wonder if we can minimise them or split them into separate PRs (if that's too complicated for this PR, it's at least something that we should keep in mind for future work).

**Comment by draganescu** [comment]
For the navigation block it says:

> Confirm each typography control works appropriately (Note: this PR does not make any specific block's typography controls "default" in terms of always displaying within the new panel)

In my testing the Appearance (thin, bold, semibold) panel has no effect and the font size does not increase font size visually (something does happen, but the font size of menu items remains the same). Maybe Appearance depends on font?

File: packages/block-editor/src/hooks/font-appearance.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/font-appearance.js:41]: The primary [typography hook needs to check](https://github.com/WordPress/gutenberg/blob/c4970287db90fee9970df8b0013b143138c20e3e/packages/block-editor/src/hooks/typography.js#L78-L84) if these controls are displayed or not to handle the `ToolsPanelItem` inclusion. This made all these checks redundant. After receiving feedback elsewhere I removed them from the individual hooks.

This is also the case in the [global styles sidebar when switched to the `ToolsPanel`.](https://github.com/WordPress/gutenberg/blob/c4970287db90fee9970df8b0013b143138c20e3e/packages/edit-site/src/components/sidebar/typography-panel.js#L109)

I can reinstate the redundant checks if you feel strongly about it.

File: packages/block-editor/src/hooks/typography.scss
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/typography.scss]: These were primarily to improve the display of the font size picker when the superfluous reset button is omitted. Ideally, we'd have the new typography components in place however I believe the current plan is for this to land before that occurs.

Would it be better to just omit these and leave the slightly ugly controls for follow-ups to fix or the new font size picker to be implemented?

File: packages/components/src/font-size-picker/index.js
Comment (aaronrobertshaw) [reply packages/components/src/font-size-picker/index.js]: > Is allowReset a prop that is also used on other similar components?

A quick search shows the `BoxControl`, `ComboboxControl`, and `RangeControl` all offer an `allowReset` prop.

> If this is a new prop, we should add documentation for it (e.g. in the README) 

If it's decided we wish to proceed with allowing for the reset button to be excluded from the font size picker, I will split that out into a separate PR and add the prop to the README along with a CHANGELOG entry.

> It would be good if these changes would happen in a separate PR 

I appreciate that is always the ideal, it can be difficult to know what changes are required and implement/test them in isolation, or manage a chain of dependant PRs. 

This mirrors the [previous discussion](https://github.com/WordPress/gutenberg/pull/34157#issuecomment-915143297) around the ToolsPanel and its use with SlotFills. The middle ground there was to split the changes out which I'm more than happy to do. This PR also pre-dates that discussion and hasn't been updated meaningfully since.

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: I have another PR in progress eliminating the use of hardcoded classnames which will also solve this particular case. When it's ready, I'll make it the base branch for this PR allowing the highlighted changes to be omitted from this PR.

File: packages/block-editor/src/hooks/typography.scss
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/typography.scss]: I've removed these styles from this PR along with the changes to opt-out of the font size picker's reset button.

| Before Removal | After Removal |
|---|---|
| Screen Shot 2021-10-07 at 3 23 20 pm | Screen Shot 2021-10-07 at 3 36 59 pm |

File: packages/components/src/font-size-picker/index.js
Comment (aaronrobertshaw) [reply packages/components/src/font-size-picker/index.js]: I've removed these `FontSizePicker` changes and will create a separate PR for them tomorrow.

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: These style changes have been removed as they are addressed by https://github.com/WordPress/gutenberg/pull/35415.

**Comment by aaronrobertshaw** [comment]
> In my testing the Appearance (thin, bold, semibold) panel has no effect and the font size does not increase font size visually (something does happen, but the font size of menu items remains the same). Maybe Appearance depends on font?

Thanks for testing @draganescu 👍 

You are correct, the font appearance is dependent upon the font and therefore potentially theme dependent as well. The TT1 Blocks theme is an easy option if you wish to see font-weight changes in action. 

https://user-images.githubusercontent.com/60436221/136334006-8d5081e0-1b92-4fbd-b16e-7535f00a9e8a.mp4

I've retested this before and after the latest changes. The font style portion of the appearance control always had the expected effect. If the font doesn't have the different weights, I'm not sure what the best approach is there. Given that is the current behaviour of the block supports, I'd suggest if it's an issue it can be addressed in a follow-up as it isn't directly related to the switch to the ToolsPanel in this PR.

I couldn't replicate the issue you reported regarding font sizes though. Any chance you could provide some extra details?

https://user-images.githubusercontent.com/60436221/136334056-5210add3-7e4f-4ee5-b986-9f4412a1c5d2.mp4

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: Awesome, thank you!

File: packages/components/src/font-size-picker/index.js
Comment (ciampo) [reply packages/components/src/font-size-picker/index.js]: > If it's decided we wish to proceed with allowing for the reset button to be excluded from the font size picker, I will split that out into a separate PR and add the prop to the README along with a CHANGELOG entry.

Sounds great 👌 

> I appreciate that is always the ideal, it can be difficult to know what changes are required and implement/test them in isolation, or manage a chain of dependant PRs.
>
> This mirrors the previous discussion around the ToolsPanel and its use with SlotFills. The middle ground there was to split the changes out which I'm more than happy to do. This PR also pre-dates that discussion and hasn't been updated meaningfully since.

You explained the situation perfectly, I agree with everything here.

> I've removed these FontSizePicker changes and will create a separate PR for them tomorrow.

Thank you!

File: packages/block-editor/src/hooks/typography.scss
Comment (ciampo) [reply packages/block-editor/src/hooks/typography.scss]: > Would it be better to just omit these and leave the slightly ugly controls for follow-ups to fix or the new font size picker to be implemented?

I think that at this point this may be a better choice.

> I've removed these styles from this PR along with the changes to opt-out of the font size picker's reset button.

Thank you! We can tackle these separately and decide what's the best strategy.

File: packages/block-editor/src/hooks/font-appearance.js
Comment (ciampo) [reply packages/block-editor/src/hooks/font-appearance.js:41]: > I can reinstate the redundant checks if you feel strongly about it.

Not at all, I just lacked context. Thank you for the explanation!

File: packages/block-editor/src/hooks/typography.scss
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/typography.scss]: I've created a [follow-up PR ](https://github.com/WordPress/gutenberg/pull/35451)for tweaking the font size and line-height controls for neater display within the `ToolsPanel`.  

At present, its base branch is this PR for ease of testing however it can just as easily be based of trunk and merged before this if needed.

https://github.com/WordPress/gutenberg/pull/35451

The new PR has been added to the main tracking issue as well.

**Comment by ciampo** [comment]
All of the feedback that I shared previously (specifically about the `@wordpress/components` package) has been addressed.

I'll let the other reviewers give a final review/approval, as they may have more context of the overall changes that this PR introduces.

**Comment by aaronrobertshaw** [comment]
With the progress being made on the individual typography controls ( e.g. https://github.com/WordPress/gutenberg/pull/35395 ), holding off on merging this switch of panel might provide the smoothest transition and remove the need for landing https://github.com/WordPress/gutenberg/pull/35451 as an interim polishing measure. 

Waiting shouldn't impact our ability to land this for 5.9.

File: packages/block-editor/src/hooks/typography.js
Comment (ramonjd) [inline packages/block-editor/src/hooks/typography.js:152]: Just checking, is this how we should set ToolsPanels defaults for style and weight going forward? 

```json
		"typography": {
			"fontSize": true,
			"lineHeight": true,
			"__experimentalFontStyle": true,
			"__experimentalFontWeight": true,
			"__experimentalDefaultControls": {
				"fontAppearance": true, 
			}
		}

```

My first instinct was to add the same key to the defaults object:

```json
		"typography": {
			"fontSize": true,
			"lineHeight": true,
			"__experimentalFontStyle": true,
			"__experimentalFontWeight": true,
			"__experimentalDefaultControls": {
				"__experimentalFontStyle": true, 
                                 "__experimentalFontWeight": true,
			}
		}

```

Would it make sense to keep the interface consistent, and check for either fontStyle or fontWeight here, and filter out the entries in the `FontAppearanceControl` accordingly? Happy to help on that if so. 

I couldn't find any background info on it apart from that we separated opt-in for font style and weight options in https://github.com/WordPress/gutenberg/pull/26844

Comment (ramonjd) [reply packages/block-editor/src/hooks/typography.js:152]: I spoke a bit too prematurely sorry.

I _think_ I see why a single default flag has to be here now, for the case where we opt-in to both, but only try show one by default: that won't work with the combined dropdown. 

Well, it might, but it seems overly complicated to keep adding/removing dropdown items.

Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/typography.js:152]: Yeah, the dual-purpose control complicates things a bit in that it doesn't follow the assumed 1:1 pattern. 

It was a while ago I made the call here but I believe my thinking, in the end, was to keep it as simple as possible. The `__experimentalDefaultControls` relates to which controls are shown by default, not which "features". So what we are really making a call on here is whether the font appearance control shows or not. In that sense I'd say the current flag fits.

I'm certainly open to other ideas if better ones present themselves.

Comment (ramonjd) [reply packages/block-editor/src/hooks/typography.js:152]: > The __experimentalDefaultControls relates to which controls are shown by default, not which "features".

This makes sense 👍 Thanks for the follow up explanation.

**Comment by aaronrobertshaw** [comment]
I've rebased this PR again.

In addition, I've removed the change of the Global Styles typography UI to the `ToolsPanel`. The `ToolsPanel` doesn't fit with the new navigation in the Global Styles sidebar. I think any desired changes to the Global Styles sidebar would be better addressed in a separate PR.

The following screenshot is how the Global Styles panel appeared before the removal of those changes as described above:

Screen Shot 2021-10-20 at 7 57 48 pm

**Comment by aaronrobertshaw** [comment]
This has now been updated to handle the font-family moving from a property on the style attribute to a top level attribute itself. The font-size reset e2es have also been updated to be `ToolsPanel` aware again.

**Review by apeatling — APPROVED** [review]
I've gone through a number of different blocks and tried all typography settings on, off, and with them reset. Everything worked as expected. I can't see a reason for this not to be approved, so I'll kick the process off. :)

There are a couple of UI issues that are related to components that I don't think need to be fixed prior to this being merged. I can follow up with those afterwards.

**Comment by ramonjd** [comment]
I've been testing this PR as a base for [Typography block support: add typography support and defaults #34064](https://github.com/WordPress/gutenberg/pull/34064).

The ToolsPanel is working as expected for all the blocks I'm testing. 

Here's the Site Title, which opts-in to all Typography support:

Oct-22-2021 11-59-40

Looks great!

I think it's also good to go! It'll also undergo more testing as a product of testing https://github.com/WordPress/gutenberg/pull/34064.

**Comment by aaronrobertshaw** [comment]
Thanks for the testing @apeatling and @ramonjd. The more eyes on this one the better I think.

Having said that, this PR is probably in a good place to push forward with a view to merging soon. 

Because this PR touches a lot of files, I'll write up a bit of a summary to help break down the changes and make it easier for other reviewers or reference in the future.

**Comment by aaronrobertshaw** [comment]
## Summary
### New InspectorControls slot
As part of the switch to the ToolsPanel, this PR adds a [new Slot](https://github.com/WordPress/gutenberg/pull/33744/files#diff-068538657061c47f0a58cfc331259bbeabe11bcc0c576a0cc496d262e5b0038c) to the [grouped Inspector Controls](https://github.com/WordPress/gutenberg/pull/33744/files#diff-bf2b0d4298b54759560099bf9fdcc7fea503c0996942e96840a53f3dc9ccaa1d). 

This allows for the injection of ad hoc controls by blocks or plugins into a block support’s panel. This way we can avoid overlapping panel responsibilities. An example of this could be a block adding a drop cap control into the Typography panel even though that is not currently offered through block supports. 

### Updating block support hooks

Each typography block support feature requires its hook to be updated with util functions to facilitate resetting or checking the presence of values. The block support edit components have also had their [checks that skip rendering streamlined](https://github.com/WordPress/gutenberg/pull/33744/files#diff-d8545a5d8f43655064236f2ea22b7447685448ccc01ca78f1395858bc35a0a63L39-L41). These will now only be made once, at the `ToolsPanel` level.

`has<Feature>Value`  

This will check whether values have been set in a block’s attributes. Depending on the feature, it might be top level attributes or a property on the style attribute that must be checked.

`reset<Feature>`

This handles clearing values from a block’s attributes and will also vary per feature.

#### Updated Hooks
* [Font Appearance](https://github.com/WordPress/gutenberg/pull/33744/files?authenticity_token=WxIIx55M1QrlrnNwImMp%2FAL9%2Bvs2LHTgS3YgqpaKG%2FXBDIcQ2WAZ%2BNaRD3pNCKHMmFiGkpusJHTzMgwEvvnQ1w%3D%3D&hide-deleted-files=true&file-filters%5B%5D=#diff-d8545a5d8f43655064236f2ea22b7447685448ccc01ca78f1395858bc35a0a63)
* [Font Family](https://github.com/WordPress/gutenberg/pull/33744/files?authenticity_token=WxIIx55M1QrlrnNwImMp%2FAL9%2Bvs2LHTgS3YgqpaKG%2FXBDIcQ2WAZ%2BNaRD3pNCKHMmFiGkpusJHTzMgwEvvnQ1w%3D%3D&hide-deleted-files=true&file-filters%5B%5D=#diff-c3545e3da7f8b8905fa1e7091138ad75ff6f452302b60fd4daa65f722dc87972)
* [Font Size](https://github.com/WordPress/gutenberg/pull/33744/files?authenticity_token=WxIIx55M1QrlrnNwImMp%2FAL9%2Bvs2LHTgS3YgqpaKG%2FXBDIcQ2WAZ%2BNaRD3pNCKHMmFiGkpusJHTzMgwEvvnQ1w%3D%3D&hide-deleted-files=true&file-filters%5B%5D=#diff-d75baee3a283000ab0137c7d8d345df855c8335a29cc1528b46b5fba51421694)
* [Letter Spacing](https://github.com/WordPress/gutenberg/pull/33744/files?authenticity_token=WxIIx55M1QrlrnNwImMp%2FAL9%2Bvs2LHTgS3YgqpaKG%2FXBDIcQ2WAZ%2BNaRD3pNCKHMmFiGkpusJHTzMgwEvvnQ1w%3D%3D&hide-deleted-files=true&file-filters%5B%5D=#diff-7f6873e89e742b9f974a2b72f19ac8dc16b704d513aa0c2d31e29be331519c95)
* [Line Height](https://github.com/WordPress/gutenberg/pull/33744/files?authenticity_token=WxIIx55M1QrlrnNwImMp%2FAL9%2Bvs2LHTgS3YgqpaKG%2FXBDIcQ2WAZ%2BNaRD3pNCKHMmFiGkpusJHTzMgwEvvnQ1w%3D%3D&hide-deleted-files=true&file-filters%5B%5D=#diff-8988607ccc9e18450e63ac42705a4455097367b84f10d065682be0873303d29f)
* [Text Decoration](https://github.com/WordPress/gutenberg/pull/33744/files?authenticity_token=WxIIx55M1QrlrnNwImMp%2FAL9%2Bvs2LHTgS3YgqpaKG%2FXBDIcQ2WAZ%2BNaRD3pNCKHMmFiGkpusJHTzMgwEvvnQ1w%3D%3D&hide-deleted-files=true&file-filters%5B%5D=#diff-14b6eaf2230374fcfcbf151f336759f24d235d1b5602cfaaf54940557de430a4)
* [Text Transform
](https://github.com/WordPress/gutenberg/pull/33744/files?authenticity_token=WxIIx55M1QrlrnNwImMp%2FAL9%2Bvs2LHTgS3YgqpaKG%2FXBDIcQ2WAZ%2BNaRD3pNCKHMmFiGkpusJHTzMgwEvvnQ1w%3D%3D&hide-deleted-files=true&file-filters%5B%5D=#diff-eacf1767d7a9726928d265bd5fcaf9b4525352ce923b71959c35da9433fe1787)

### Switching to the ToolsPanel
The real shift to using the `ToolsPanel` occurs within the [`typography.js` hook](https://github.com/WordPress/gutenberg/pull/33744/files#diff-ad62fba616df95e068cc14d4079cdaa9f4f774b429a04f42ca6749cab6d6baea).

Here, the original use of a `PanelBody` that is rendered into the general `InspectorControls` slot, is replaced. Each feature or control that would have been rendered in the `PanelBody` is wrapped in a new `ToolsPanelItem` component and rendered into the new `typography` grouped Inspector Controls slot.  The main `typography` hook also [creates functions](https://github.com/WordPress/gutenberg/pull/33744/files?authenticity_token=WxIIx55M1QrlrnNwImMp%2FAL9%2Bvs2LHTgS3YgqpaKG%2FXBDIcQ2WAZ%2BNaRD3pNCKHMmFiGkpusJHTzMgwEvvnQ1w%3D%3D&file-filters%5B%5D=.js&hide-deleted-files=true#diff-ad62fba616df95e068cc14d4079cdaa9f4f774b429a04f42ca6749cab6d6baeaR96-R105) that are used by the `ToolsPanel` to carry out resets on those individual items when the corresponding menu item is toggled on or off.

### Styling
Not all components used in block support controls have been converted to the new hook and context system. As such, we need to rely on [some additional CSS](https://github.com/WordPress/gutenberg/pull/33744/files#diff-d4cf282689acbda87bcf61ee865b3a15601b5fa359b1829afa68b53778f66cb0) to polish up the display of these controls within the `ToolsPanel` context. These styles will eventually be removed as the components used, adopt the context system.

### Typography support opt-in & default controls
It was requested that the paragraph block opts into the typography support for at least font size and line height and these controls be displayed by default within the `ToolsPanel`.  This is the only block that has its support changed in this PR. All other changes in terms of opt-in or defaults occur in a follow-up PR to this one: https://github.com/WordPress/gutenberg/pull/34064

### E2E test changes
As a result of the switch to the `ToolsPanel`, the font size picker will now opt out of displaying the reset button as it's superfluous to the reset functionality offered by the `ToolsPanel` menu. This affected the e2e’s in that they previously relied on tabbing to the reset button to ensure a font size could be reset. This was [updated to work via the menu](https://github.com/WordPress/gutenberg/pull/33744/files?authenticity_token=SgJGxW2oliDE0UlmwUF9wiEWjBzPr2F3%2BANib4jxpl3QHMkSKoRa0vfuNWyuKvXyu7PwdWIvMeNAR07BoIJtfw%3D%3D&file-filters%5B%5D=.js&hide-deleted-files=true#diff-35619ba21f6e88dede90f9e2f6dbba4265c614ec45014b621a8593da4d73e995).

### Clean up
There is currently no use of the combined [text decoration](https://github.com/WordPress/gutenberg/pull/33744/files#diff-f29b18b714ab7d48c27c70e1f143befc4cf2096c37997b5ebb6fcd00aba56aa4) and [transform](https://github.com/WordPress/gutenberg/pull/33744/files#diff-cfb9ea5c3e49543ea4328f4b9fd80e124a1ea772660bbf98cbd22036ccffaed3) controls. These are used individually. Moving forward in the `ToolsPanel` these will have to remain individually used to allow for resetting each control via the panel menu. As such those individual controls are now [exported through the block-editor](https://github.com/WordPress/gutenberg/pull/33744/files#diff-d7fadfb5b3a2caa9f5196dac0a9d3094f6bad11a78b2bd83f366887463602202) components package.

File: packages/block-editor/src/hooks/typography.js
Comment (andrewserong) [inline packages/block-editor/src/hooks/typography.js:150]: In [font-appearance-control](https://github.com/wordpress/gutenberg/blob/8d8e07699df7c31fd89c07406356fb779de90b33/packages/block-editor/src/components/font-appearance-control/index.js#L156) the label is conditional as to whether font styles, font weights, or both are available. Should we add a conditional here, too? (or abstract that logic into its own function so it can be used in both places?

I noticed while testing the Heading block in TT1-Blocks that it felt a little confusing the label in the menu being different to the one in the panel item itself.

| Panel label | Menu label |
| --- | --- |
| image | image |

**Review by andrewserong — APPROVED** [review]
Great work @aaronrobertshaw, it's super satisfying seeing the ToolsPanel rolled out to the Typography support, and I like the consistent patterns used for all the different hooks 👍

Each of the typography supports is working well in my testing (along with reset and rest all behaviour). I noticed with the Navigation block that setting text decoration updates the setting correctly, and the styling on the front end renders correctly, but the style doesn't appear to win the specificity/inheritance within the editor if we're using the _editor_ version of the Navigation block (e.g. clicking Edit on the block so we're not using the server rendered view in the editor). This is the same behaviour as on trunk, though, so unrelated to this PR and not a blocker.

The code change is looking good, I noticed one issue with the Font Appearance control. We might want to update the `label` in the menu dropdown so that it shares the same logic as the label used by the hook (I've left a comment about this).

Since this is already a fairly big PR, I'd be in support of merging this in as-is, and us looking at tweaking the label for Font Appearance in a quick follow-up PR, since we're already going to be doing a bunch of other follow-ups anyway.

LGTM! 🚀

**Comment by aaronrobertshaw** [comment]
Thanks for the thorough review @andrewserong, always appreciate it 👍 

>  I noticed with the Navigation block that setting text decoration updates the setting correctly, and the styling on the front end renders correctly, but the style doesn't appear to win the specificity/inheritance within the editor if we're using the editor version of the Navigation block (e.g. clicking Edit on the block so we're not using the server rendered view in the editor).

I've been able to replicate this but only after selecting the individual item/link within the Navigation block and choosing Edit from the block's toolbar. After a little exploration, it appears that we end up with a div inside the item's `<a>` and it has a `text-decoration-line` style that interferes with the `text-decoration` inheritance.

As mentioned, it occurs on `trunk` and is not directly related to this PR. I'll happily address it in a follow-up.

> I noticed one issue with the Font Appearance control. We might want to update the label in the menu dropdown so that it shares the same logic as the label used by the hook 

Nice catch! 

I like the idea of addressing this in a follow-up as well. Given the use of the label as a key for the panel's internal menu item state and for item registration, I'll need to check for any edge cases around making this change when used in conjunction with the typography Inspector Controls SlotFill.

---

Thank you everyone for the testing and reviews. As we have multiple approvals and plans for a few small follow ups I'll merge this now to unblock https://github.com/WordPress/gutenberg/pull/34064

**Comment by aaronrobertshaw** [comment]
A potential fix for the Nav item not inheriting the text-decoration style correctly is up in https://github.com/WordPress/gutenberg/pull/35859

**Pull request merged**
